atoma-proxy-service/src/components/grafana.rs

```diff
@@ -0,0 +1,207 @@
+            api_token,
+            dashboard_tag,
+        }
+    }
+
+    /// Get the UIDs of all dashboards with the specified tag
+    ///
+    /// # Returns
+    ///
+    /// A vector of dashboard UIDs
+    #[instrument(level = "info", skip_all)]
+    pub async fn get_dashboard_uids(&self) -> Result<Vec<String>, GrafanaError> {
+        let client = Client::new();
+        let request_url = format!("{}/api/search?tag={}", self.url, self.dashboard_tag);
+        let response = client
+            .get(&request_url)
+            .header("Accept", "application/json")
+            .header("Content-Type", "application/json")
+            .header("Authorization", format!("Bearer {}", self.api_token))
+            .send()
+            .await?;
+
+        if response.status().is_success() {
+            let dashboards = response.json::<Vec<DashboardList>>().await?;
+            Ok(dashboards
+                .into_iter()
+                .map(|dashboard| dashboard.uid)
+                .collect())
+        } else {
+            Err(GrafanaError::FailedRequest(
+                response.error_for_status().unwrap_err(),
+            ))
+        }
+    }
+
+    /// Get a dashboard by its UID
+    ///
+    /// # Arguments
+    ///
+    /// * `dashboard_uid` - The UID of the dashboard to get
+    ///
+    /// # Returns
+    ///
+    /// The dashboard with the specified UID
+    #[instrument(level = "info", skip(self))]
+    pub async fn get_dashboard(&self, dashboard_uid: String) -> Result<Dashboard, GrafanaError> {
+        let client = Client::new();
+        let request_url = format!("{}/api/dashboards/uid/{}", self.url, dashboard_uid);
+        let response = client
+            .get(&request_url)
+            .header("Accept", "application/json")
+            .header("Authorization", format!("Bearer {}", self.api_token))
+            .send()
+            .await?;
+
+        if response.status().is_success() {
+            let dashboard = response.json::<Dashboard>().await?;
+            Ok(dashboard)
+        } else {
+            Err(GrafanaError::FailedRequest(
+                response.error_for_status().unwrap_err(),
+            ))
+        }
+    }
+
+    /// Query data from a Grafana dashboard
+    ///
+    /// # Arguments
+    ///
+    /// * `dashboard` - The dashboard to query data from
+    ///
+    /// # Returns
+    ///
+    /// A vector of tuples containing the title of the panel and the data
+    #[instrument(level = "info", skip_all)]
+    pub async fn query_data(
+        &self,
+        dashboard: Dashboard,
+    ) -> Result<Vec<(String, Value)>, GrafanaError> {
+        let client = Client::new();
+        let request_url = format!("{}/api/ds/query", self.url);
+        let queries: Vec<(String, Query)> = dashboard.into();
+        let mut results = Vec::new();
+        for (title, query) in queries {
+            let response = client
+                .post(&request_url)
+                .header("Accept", "application/json")
```

Comment:
There's some repetition in adding a header to each request, this could be extracted into a helper method

e.g.

```rust
impl Grafana {
    fn prepare_request(&self, builder: reqwest::RequestBuilder) -> reqwest::RequestBuilder {
        builder
            .header("Accept", "application/json")
            .header("Authorization", format!("Bearer {}", self.api_token))
            .header("Content-Type", "application/json")
    }
}
```

Reply:
Or process all headers at once